### PR TITLE
[tagger] return copies of the cache to avoid modifications

### PR DESCRIPTION
**Pull request by xvello**
For performance reasons (DSD origin detection), the tagger returns a `[]string` slice that is cached and reused. Coupled to golang's un-intuitive `append` behaviour, this led to the docker corecheck creeping tags up to the tagstore cache.

[Playground link](https://play.golang.org/p/dFRiZ0QR7m-) illustrating the issue and one possible fix.

### Symptoms

- one tag missing from AD metrics, replaced by a `docker_network` tag
- duplicate contexts in docker / kubernetes metrics, leading to `sum` aggregation errors

### What does this PR do?

Make the tagger return copies of the slices to avoid users changing the common cache.

### Backport

The `datadog/agent-dev:backport-1574` backport image ships this fix on top of 6.1.2, while we work on a bugfix release.

### CODE REVIEW

File: pkg/logs/input/container/docker.go [hunk 1]
```diff
@@ -226,7 +227,9 @@ func (dt *DockerTailer) checkForNewDockerTags() {
 		log.Warn(err)
 	} else {
 		if !reflect.DeepEqual(tags, dt.containerTags) {
-			dt.containerTags = tags
+			// Copy the slice to avoid side effects when append-ing the other tags
+			dt.containerTags = make([]string, len(tags))
+			copy(dt.containerTags, tags)
```
Comment (mfpierre) [inline pkg/logs/input/container/docker.go]: not sure why we're copying there, is it just a safety measure? we don't do it in the core check

Comment (xvello) [reply pkg/logs/input/container/docker.go]: As the slice is passed down and `append`ed later, see https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/message/origin.go#L29-L40 , I went with the safest route, as:

- it is less intrusive than patching the rest of the downstream processing
- we are behind a caching later (`if !reflect.DeepEqual(tags, dt.containerTags)`)

**Comment by mfpierre** [comment]
As discussed, it's a tradeoff between safety and performance but I'm leaning toward safety and have the tagger making the copy itself to be secured by default because it's gonna be hard to enforce the `append` rule in the future

**Review by ajacquemot — DISMISSED** [review]

**Comment by CharlyF** [comment]
Good catch! I have some feedback and a question as well:
```
- do NOT do `tags := append(tags_from_tagger, my_tags...)`
- instead, do `tags := append(my_tags, tags_from_tagger...)`
```
Is error prone I think, if we go this route, I suggest we log a user facing error or something that surfaces that some tags got mixed up.
 
Then, I'm not sure if I understand @mfpierre's comment correctly: Are you suggesting appending the docker tag slice to a slice owned by the tagger (similarly to the AD resolver and dsd) ? If so, I understand the perf trade off and I think it'd be interesting to see how bad it'd be. I'm just thinking about troubleshooting an error sourcing from this in a year, and it might not be easy so it might be worth it to impact the perfs.

**Comment by mfpierre** [comment]
@CharlyF having the tagger returning a copy of the slice by default and having an unsecure method to return directly the slice of the tagger in case of critical performance code path (but I'm actually curious to see if we really need to do the latest, ie: is the perf overhead really worth it?)

File: pkg/tagger/README.md [hunk 2]
```diff
@@ -17,6 +17,12 @@ The tagger is also available to python checks via the `tagger` module exporting
 the `get_tags()` function. This function accepts the same arguments as the Go `Tag()`
 function, and returns an empty list on errors.
 
+**Note:** for performance reasons, the slices returned by the tagger are cached and
```
Comment (hkaj) [inline pkg/tagger/README.md]: I don't think the performance required it warrants the risk. The tagger should expose a safe API. If we need it to be fast in specific cases I'm okay with @mfpierre 's proposal of having an explicit unsafe method, which we'll introduce if we need to, but we probably don't right now.

**Review by mfpierre — APPROVED** [review]
Good job spotting this 💯 
LGTM :shipit:

**Pull request merged**